### PR TITLE
net/socket: add SO_SNDBUF support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/socket: add SO_SNDBUF support
net/wrbuffer: add tcp/udp_wrbuffer_inqueue_size() interface

Reference here:
https://www.freebsd.org/cgi/man.cgi?query=setsockopt&sektion=2&format=html

## Impact

add SO_SNDBUF support to limit the the sharing of iob buffers usage under multiple TCP/UDP connections 

## Testing

iperf client with multiple instance to run the net performance at same time

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

File: net/tcp/tcp_wrbuffer.c [hunk 1]
```diff
@@ -152,17 +152,31 @@ uint32_t tcp_wrbuffer_inqueue_size(FAR struct tcp_conn_s *conn)
...
+#if CONFIG_NET_SEND_BUFSIZE > 0
+  if (conn)
+    {
+      /* If the send buffer size exceeds the send limit, wait for the write
+       * buffer to be released
+       */
+
+      while (tcp_wrbuffer_inqueue_size(conn) > conn->snd_bufs)
+        {
+          net_lockedwait_uninterruptible(&conn->snd_sem);
```
Comment (yamt) [inline net/tcp/tcp_wrbuffer.c:175]: i guess this kind of things should be interruptible. but it isn't a new problem in this change.

File: net/tcp/tcp_wrbuffer.c [hunk 2]
```diff
@@ -105,6 +105,44 @@ void tcp_wrbuffer_initialize(void)
...
+      for (entry = sq_peek(&conn->unacked_q); entry; entry = sq_next(entry))
+        {
+          wrb = (FAR struct tcp_wrbuffer_s *)entry;
+          total += TCP_WBSENT(wrb);
+        }
+
+      for (entry = sq_peek(&conn->write_q); entry; entry = sq_next(entry))
+        {
+          wrb = (FAR struct tcp_wrbuffer_s *)entry;
+          total += TCP_WBSENT(wrb);
```
Comment (yamt) [inline net/tcp/tcp_wrbuffer.c:139]: TCP_WBSENT is on-wire bytes, which is not appropriate thing to look at for SO_SNDBUF.

i guess TCP_WBPKTLEN is more appropriate. how do you think?

**Review by yamt — CHANGES_REQUESTED** [review]
i looked at the tcp part.

* tcp_wrbuffer_tryalloc is not the only place we allocate IOBs. we allocate them in TCP_WBTRYCOPYIN too. i suspect it might be simpler to perform the check in psock_tcp_send.
* i guess you need to update poll.

File: net/Kconfig [hunk 3]
```diff
@@ -123,6 +123,12 @@ config NET_RECV_BUFSIZE
 	---help---
 		This is the default value for receive buffer size.
 
+config NET_SEND_BUFSIZE
```
Comment (yamt) [inline net/Kconfig:126]: maybe depends on NET_TCP_WRITE_BUFFERS?

**Comment by anchao** [comment]
> i looked at the tcp part.
> 
> * tcp_wrbuffer_tryalloc is not the only place we allocate IOBs. we allocate them in TCP_WBTRYCOPYIN too. i suspect it might be simpler to perform the check in psock_tcp_send.
> * i guess you need to update poll.

@yamt san,
Think for a while, psock_tcp_send not suitable to check the buffer size, because it's hard to find a suitable place to notify that there are available buffers, TCP_WBTRYCOPYIN is a potential risk, but every time of tcp send() will only do the iobcopyin once, the next send() will be check further by tcp_wrbuffer_tryalloc(), I don't think it's a serious scene

File: net/tcp/tcp_wrbuffer.c (hunk 2 above)
Comment (anchao) [reply net/tcp/tcp_wrbuffer.c:139]: Done

File: net/Kconfig (hunk 3 above)
Comment (anchao) [reply net/Kconfig:126]: Done

**Comment by yamt** [comment]
> > i looked at the tcp part.
> > 
> > * tcp_wrbuffer_tryalloc is not the only place we allocate IOBs. we allocate them in TCP_WBTRYCOPYIN too. i suspect it might be simpler to perform the check in psock_tcp_send.
> > * i guess you need to update poll.
> 
> @yamt san,
> Think for a while, psock_tcp_send not suitable to check the buffer size, because it's hard to find a suitable place to notify that there are available buffers,

do you mean where to put the semaphore or an equivalent of it?
what's wrong with eg. conn->snd_buf_sem?
psock_send_eventhandler can wake it up when some buffers are acked by the peer.

> TCP_WBTRYCOPYIN is a potential risk, but every time of tcp send() will only do the iobcopyin once, the next send() will be check further by tcp_wrbuffer_tryalloc(), I don't think it's a serious scene

maybe it works for the majority of cases, yes.
i suspect it can be a problem for a small sndbuf though.

**Comment by anchao** [comment]
> do you mean where to put the semaphore or an equivalent of it?
> what's wrong with eg. conn->snd_buf_sem?
> psock_send_eventhandler can wake it up when some buffers are acked by the peer.

not just psock_send_eventhandler(). I think semaphore put should be done wherever wrbuffer is released.

```
net/tcp/tcp_send_buffered.c:214:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
net/tcp/tcp_send_buffered.c:220:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
net/tcp/tcp_send_buffered.c:434:                  tcp_wrbuffer_release(wrb);
net/tcp/tcp_send_buffered.c:607:              tcp_wrbuffer_release(wrb);
net/tcp/tcp_send_buffered.c:672:              tcp_wrbuffer_release(wrb);
net/tcp/tcp_send_buffered.c:1246:  tcp_wrbuffer_release(wrb);
net/tcp/tcp_conn.c:741:      tcp_wrbuffer_release(wrbuffer);
net/tcp/tcp_conn.c:747:      tcp_wrbuffer_release(wrbuffer);
net/tcp/tcp_wrbuffer.c:157:      tcp_wrbuffer_release(wrb);
net/tcp/tcp_wrbuffer.c:213:      tcp_wrbuffer_release(wrb);
```

> maybe it works for the majority of cases, yes.
> i suspect it can be a problem for a small sndbuf though.

This issue is inherent in the configuration of small buffers, SO_SNDBUF is just a suggestion value, How about keeping the current behavior?

**Comment by yamt** [comment]
> > do you mean where to put the semaphore or an equivalent of it?
> > what's wrong with eg. conn->snd_buf_sem?
> > psock_send_eventhandler can wake it up when some buffers are acked by the peer.
> 
> not just psock_send_eventhandler(). I think semaphore put should be done wherever wrbuffer is released.

as far as i understand, the data in the send buffer is disposed only when

* acked by the peer (psock_send_eventhandler)
* connection is closed

> 
> ```
> net/tcp/tcp_send_buffered.c:214:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
> net/tcp/tcp_send_buffered.c:220:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
> net/tcp/tcp_send_buffered.c:434:                  tcp_wrbuffer_release(wrb);
> net/tcp/tcp_send_buffered.c:607:              tcp_wrbuffer_release(wrb);
> net/tcp/tcp_send_buffered.c:672:              tcp_wrbuffer_release(wrb);
> net/tcp/tcp_send_buffered.c:1246:  tcp_wrbuffer_release(wrb);
> net/tcp/tcp_conn.c:741:      tcp_wrbuffer_release(wrbuffer);
> net/tcp/tcp_conn.c:747:      tcp_wrbuffer_release(wrbuffer);
> net/tcp/tcp_wrbuffer.c:157:      tcp_wrbuffer_release(wrb);
> net/tcp/tcp_wrbuffer.c:213:      tcp_wrbuffer_release(wrb);
> ```
> 
> > maybe it works for the majority of cases, yes.
> > i suspect it can be a problem for a small sndbuf though.
> 
> This issue is inherent in the configuration of small buffers, SO_SNDBUF is just a suggestion value, How about keeping the current behavior?

i'm not sure what you mean here. can you explain a bit?

**Comment by anchao** [comment]
> i'm not sure what you mean here. can you explain a bit?

If we move the check outside the wrbuffer, we will duplicate a lot of sem_post codes in these places. Do you want to do this? I don't think there is any place like net/tcp/tcp_wrbuffer.c is more suitable.

```
net/tcp/tcp_send_buffered.c:214:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
net/tcp/tcp_send_buffered.c:220:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
net/tcp/tcp_send_buffered.c:434:                  tcp_wrbuffer_release(wrb);
net/tcp/tcp_send_buffered.c:607:              tcp_wrbuffer_release(wrb);
net/tcp/tcp_send_buffered.c:672:              tcp_wrbuffer_release(wrb);
net/tcp/tcp_send_buffered.c:1246:  tcp_wrbuffer_release(wrb);
net/tcp/tcp_conn.c:741:      tcp_wrbuffer_release(wrbuffer);
net/tcp/tcp_conn.c:747:      tcp_wrbuffer_release(wrbuffer);
net/tcp/tcp_wrbuffer.c:157:      tcp_wrbuffer_release(wrb);
net/tcp/tcp_wrbuffer.c:213:      tcp_wrbuffer_release(wrb);
```
like this ? 

net/tcp/tcp_wrbuffer.c:

```
tcp_wrbuffer_release(wrb);
#if CONFIG_NET_SEND_BUFSIZE > 0
  if (conn)
    {
      int val = 0;

      nxsem_get_value(&conn->snd_sem, &val);
      if (val < 0)
        {
          nxsem_post(&conn->snd_sem);
        }
    }
#endif
```

net/tcp/tcp_conn.c:
```
tcp_wrbuffer_release(wrb);
#if CONFIG_NET_SEND_BUFSIZE > 0
  if (conn)
    {
      int val = 0;

      nxsem_get_value(&conn->snd_sem, &val);
      if (val < 0)
        {
          nxsem_post(&conn->snd_sem);
        }
    }
#endif
```

**Comment by xiaoxiang781216** [comment]
Let's add the wait in psock_tcp_send and notify in the done callback directly?

**Comment by yamt** [comment]
> > i'm not sure what you mean here. can you explain a bit?
> 
> If we move the check outside the wrbuffer, we will duplicate a lot of sem_post codes in these places. Do you want to do this? I don't think there is any place like net/tcp/tcp_wrbuffer.c is more suitable.

you need to signal waiters for TCP_WBTRIM too.
to me it looks simpler to do it in psock_send_eventhandler directly.
(two places in the function, where `TCP_SEQ_GT(ackno, TCP_WBSEQNO(wrb)` is checked)

> 
> ```
> net/tcp/tcp_send_buffered.c:214:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
> net/tcp/tcp_send_buffered.c:220:          tcp_wrbuffer_release((FAR struct tcp_wrbuffer_s *)entry);
> net/tcp/tcp_send_buffered.c:434:                  tcp_wrbuffer_release(wrb);
> net/tcp/tcp_send_buffered.c:607:              tcp_wrbuffer_release(wrb);
> net/tcp/tcp_send_buffered.c:672:              tcp_wrbuffer_release(wrb);
> net/tcp/tcp_send_buffered.c:1246:  tcp_wrbuffer_release(wrb);
> net/tcp/tcp_conn.c:741:      tcp_wrbuffer_release(wrbuffer);
> net/tcp/tcp_conn.c:747:      tcp_wrbuffer_release(wrbuffer);
> net/tcp/tcp_wrbuffer.c:157:      tcp_wrbuffer_release(wrb);
> net/tcp/tcp_wrbuffer.c:213:      tcp_wrbuffer_release(wrb);
> ```
> 
> like this ?

yes. but see above.

> 
> net/tcp/tcp_wrbuffer.c:
> 
> ```
> tcp_wrbuffer_release(wrb);
> #if CONFIG_NET_SEND_BUFSIZE > 0
>   if (conn)
>     {
>       int val = 0;
> 
>       nxsem_get_value(&conn->snd_sem, &val);
>       if (val < 0)
>         {
>           nxsem_post(&conn->snd_sem);
>         }
>     }
> #endif
> ```
> 
> net/tcp/tcp_conn.c:
> 
> ```
> tcp_wrbuffer_release(wrb);
> #if CONFIG_NET_SEND_BUFSIZE > 0
>   if (conn)
>     {
>       int val = 0;
> 
>       nxsem_get_value(&conn->snd_sem, &val);
>       if (val < 0)
>         {
>           nxsem_post(&conn->snd_sem);
>         }
>     }
> #endif
> ```

File: net/udp/udp_sendto_buffered.c [hunk 4]
```diff
@@ -657,6 +671,23 @@ ssize_t psock_udp_sendto(FAR struct socket *psock, FAR const void *buf,
     {
       net_lock();
 
+#if CONFIG_NET_SEND_BUFSIZE > 0
+      /* If the send buffer size exceeds the send limit,
+       * wait for the write buffer to be released
+       */
+
+      while (udp_inqueue_wrb_size(conn) >= conn->sndbufs)
```
Comment (xiaoxiang781216) [inline net/udp/udp_sendto_buffered.c:679]: add + len

File: net/tcp/tcp_send_buffered.c [hunk 5]
```diff
@@ -1098,6 +1122,23 @@ ssize_t psock_tcp_send(FAR struct socket *psock, FAR const void *buf,
       psock->s_sndcb->priv  = (FAR void *)psock;
       psock->s_sndcb->event = psock_send_eventhandler;
 
+#if CONFIG_NET_SEND_BUFSIZE > 0
+      /* If the send buffer size exceeds the send limit,
+       * wait for the write buffer to be released
+       */
+
+      while (tcp_inqueue_wrb_size(conn) >= conn->snd_bufs)
```
Comment (xiaoxiang781216) [inline net/tcp/tcp_send_buffered.c:1130]: + len

**Comment by anchao** [comment]
@yamt  san, please review again

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM.

**Review by yamt — APPROVED** [review]
lgtm

**Pull request merged**